pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+Tag
+    : "tag" "(" SignedInteger ")"
+    ;
+```
+
+### Identifiers
+
+Slice supports three forms of identifiers:
+- An unscoped identifier. These consist of a letter, followed by any number of letters, digits, or underscores.
+- A relatively scoped identifier. These consist of one or more unscoped identifiers, separated by `::` tokens.
+- A globally scoped identifier. These consist of a `::` token, followed by a relatively scoped identifier.
+
+```ebnf {% showTitle=false %}
+Identifier
+    : identifier
+    ;
+
+RelativeIdentifier
+    : identifier ("::" identifier)*
+    ;
+
+GlobalIdentifier
+    : ("::" identifier)+
+    ;
+```
+
+Examples of valid identifiers:
+
+```slice
+// Examples of unscoped identifiers
+foo9
+_bar_
+BAZ
+
+// Examples of relatively scoped identifiers
+foo9
+foo9::_bar_::BAZ
+
+// Examples of globally scoped identifiers
+::foo9
+::foo9::_bar_::BAZ
+```
+
+All Slice [keywords] satisfy the definition of an unscoped identifier.
+Hence the lexer always gives priority to producing keyword tokens, even when it could produce an identifier token.
+To use a keyword as an identifier, you can escape the keyword by prefixing it with a backslash.
+For scoped identifiers, each individual identifier must be escaped this way.
+It is legal to escape an identifier that is not a keyword, though doing so has no effect.
+
+```slice
+struct   // The struct keyword
+\struct  // An unscoped identifier
+\foo     // An unscoped identifier
+
+::\struct::\module // A globally scoped identifier
+```
+
+### Integer literals
+
+Slice supports three forms of integer literals:
+- A decimal literal for base-10 numbers. These consist of one or more digits between `0` and `9`.
+- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
+- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.
+
+Additionally, underscores can appear between any two characters of an integer literal.
+These underscores carry no semantic meaning and are discarded by the parser.
+
+Examples of valid integer literals:
+
+```slice
+123
+0xFF
+0b1010
+
+335_445_996
+0x_ab_cd_ef  // Same as 0xabcded
+0b0_______1  // Same as 0b01
+
+0x_0b1101  // hexadecimal literal of value 725249
+
+// All have a value of 0
+0
+0_00_0
+0b000
+0x0
+```
+
+{% callout type="note" %}
+Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
+{% /callout %}
+
+The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
+Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.
+
+```ebnf {% showTitle=false %}
+Integer
+    : integer_literal
+    ;
+
+SignedInteger
+    : Integer
+    | "-" Integer
+    ;
+```
+
+### String literals
```

Comment:
It would be useful to mention that string literals are used only for attribute parameters.

Reply:
Agreed, I added this to the section: `String literals can only appear as arguments to an [attribute].`